[JIT] Separate TorchScript out from custom class registration

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33586 [JIT] Separate TorchScript out from custom class registration**

This makes it so we don't have libtorch<->aten circular dependencies in internal builds

Differential Revision: [D20012588](https://our.internmc.facebook.com/intern/diff/D20012588/)

Commit:
[JIT] Separate TorchScript out from custom class registration



Differential Revision: [D20012588](https://our.internmc.facebook.com/intern/diff/D20012588/)

[ghstack-poisoned]

Commit:
Update on "[JIT] Separate TorchScript out from custom class registration"


This makes it so we don't have libtorch<->aten circular dependencies in internal builds

Differential Revision: [D20012588](https://our.internmc.facebook.com/intern/diff/D20012588/)

[ghstack-poisoned]

Commit:
Update on "[JIT] Separate TorchScript out from custom class registration"


This makes it so we don't have libtorch<->aten circular dependencies in internal builds

Differential Revision: [D20012588](https://our.internmc.facebook.com/intern/diff/D20012588/)

[ghstack-poisoned]

Commit:
Update on "[JIT] Separate TorchScript out from custom class registration"


This makes it so we don't have libtorch<->aten circular dependencies in internal builds

Differential Revision: [D20012588](https://our.internmc.facebook.com/intern/diff/D20012588/)

[ghstack-poisoned]

Commit:
Update on "[JIT] Separate TorchScript out from custom class registration"


This makes it so we don't have libtorch<->aten circular dependencies in internal builds

Differential Revision: [D20012588](https://our.internmc.facebook.com/intern/diff/D20012588/)

[ghstack-poisoned]

Commit:
Update on "[JIT] Separate TorchScript out from custom class registration"


This makes it so we don't have libtorch<->aten circular dependencies in internal builds

Differential Revision: [D20012588](https://our.internmc.facebook.com/intern/diff/D20012588/)

[ghstack-poisoned]

dzhulgakov commented on torch/csrc/jit/custom_class.cpp:
```diff
@@ -21,20 +22,127 @@ bool isCustomClass(const c10::IValue& v) {
+
+void classCallback(const detail::RegisteredClassRecord& clazz) {
+  auto classTypePtr =
+      ClassType::create(c10::QualifiedName(clazz.qualClassName), classCU());
```
do we need ClassType to exist for IValue to work? not really sure how capsule type works

Commit:
Update on "[JIT] Separate TorchScript out from custom class registration"


This makes it so we don't have libtorch<->aten circular dependencies in internal builds

Differential Revision: [D20012588](https://our.internmc.facebook.com/intern/diff/D20012588/)

[ghstack-poisoned]

jamesr66a commented on torch/custom_class.h:
```diff
@@ -56,18 +51,16 @@ class class_ {
-    classCU()->register_type(classTypePtr);
+    registeredClasses()[className] = detail::RegisteredClassRecord{
+        qualClassName,
+        typeid(c10::intrusive_ptr<CurClass>).name(),
```
Does it matter? Aren't we just looking for exact matches?

Commit:
Update on "[JIT] Separate TorchScript out from custom class registration"


This makes it so we don't have libtorch<->aten circular dependencies in internal builds

Differential Revision: [D20012588](https://our.internmc.facebook.com/intern/diff/D20012588/)

[ghstack-poisoned]
